src/transformStateWithClones.js

```diff
@@ -1,13 +1,27 @@
 'use strict';
 
-/**
- * @param {Object} state
- * @param {Object[]} actions
- *
- * @return {Object[]}
- */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const currentState = Object.assign({}, state);
+  const resultArray = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    const { type, extraData, keysToRemove } = actions[i];
+
+    if (type === 'addProperties') {
+      Object.assign(currentState, extraData);
+    } else if (type === 'removeProperties') {
+      keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    } else {
+      for (const key in currentState) {
+        delete currentState[key];
+      }
+    }
+    resultArray.push(Object.assign({}, currentState));
```

Comment:
It's better to push a copy of the object to the array at the end of each loop cycle but outside of the `switch` block, after the current action is processed. This ensures that the action is completed before the state is added to the result array.